Make clear function affect the synced data, fixes #7, fixes #13, fixes #14, fixes #17

michielbdejong opened the pull request:
- [x] Needs unit tests
- [x] Rebase once https://github.com/Kinto/kinto-chrome/pull/8 is merged

Commit:
Fix area#get, fixes #14

Commit:
Deal with creation after virtual deletion, fix #13

michielbdejong commented:
r? @n1k0

n1k0 commented on test/chrome_test.js:
```diff
@@ -83,22 +95,48 @@ describe("StorageArea", function() {
 
+    it("only deletes records virtually", function(done) {
+      area.items.list({}, { includeDeleted: true }).then(arr => {
+        console.log(arr);
```
Maybe you want an assertion here

michielbdejong replied:
Oops, yes, forgot to add that.

n1k0 commented on lib/chrome.js:
```diff
@@ -88,7 +88,7 @@ export class StorageArea {
   get(keys, callback) {
     const this_items = this.items;
-    const records = [];
+    const records = {};
```
Couldn't we use a Map here?

michielbdejong replied:
No (or we would have to cast it to an `Object` later), since we need to give an ['Object with items in their key-value mappings'](https://developer.chrome.com/extensions/storage) to the callback.

n1k0 commented on lib/chrome.js:
```diff
@@ -150,16 +150,23 @@ export class StorageArea {
+      return this_items.get(id, { includeDeleted: true })
+      .then(function(old_record) {
+        return updateItem(old_record.data);
+      }, function(reason) {
```
Shouldn't we do something with the error?

michielbdejong replied:
yes, good point

michielbdejong replied:
done

n1k0 commented on lib/chrome.js:
```diff
@@ -150,16 +150,23 @@ export class StorageArea {
+        if (old_record._status === "deleted") {
+          return this_items.delete(old_record.id, { virtual: false }).then(() => {
+            return this_items.create(record, {useRecordId: true});
+          });
```
For efficiency, could we just update `old_record` with `record` data?

michielbdejong replied:
That would fail saying record not found (because it was virtually deleted)

n1k0 commented on lib/chrome.js:
```diff
@@ -150,16 +150,23 @@ export class StorageArea {
-      return this_items.get(id).then(function(old_record) {return updateItem(old_record);},
-                                     function(reason) {return createItem();});
+      return this_items.get(id, { includeDeleted: true })
+      .then(function(old_record) {
```
nit: indentation

michielbdejong replied:
fixed

n1k0 commented on lib/chrome.js:
```diff
@@ -183,7 +190,14 @@ export class StorageArea {
   clear(callback) {
-    this.items.clear().then(function() {if (callback) {callback();}});
+    this.items.list()
+    .then(records => Promise.all(records.data.map(record =>
```
nit: indentation

michielbdejong replied:
fixed

n1k0 commented on lib/chrome.js:
```diff
@@ -183,7 +190,14 @@ export class StorageArea {
-    this.items.clear().then(function() {if (callback) {callback();}});
+    this.items.list()
+    .then(records => Promise.all(records.data.map(record =>
+      this.items.delete(record.id))))
```
Reasoning for not using collection.clear?

michielbdejong replied:
We want the operation to result in all records being deleted on the server as well.

n1k0 commented on test/chrome_test.js:
```diff
@@ -70,12 +70,24 @@ describe("StorageArea", function() {
+        area.remove("something", function() {
+          area.set({"something": 2}, function() {
+            area.get(null, function(items) {
+              console.log(items);
```
We should probably remove console.log statements in tests

michielbdejong replied:
Oops, missed that one, thanks.

michielbdejong replied:
removed

Commit:
Remove all records when clear is called, fix #7

Commit:
Fix area#set, fixes #17

michielbdejong commented:
@n1k0 addressed your comments, and fixed one more issue I encountered. Each commit message references the issue it fixes, but maybe you want 4 separate PRs for this?

n1k0 commented:
Looks all good, nits here and there r+

michielbdejong commented:
Thanks! Additional comments addressed, merging.
